plugins/intel-gsc/fu-igsc-device.c

```diff
@@ -0,0 +1,844 @@
+		if (cp_mode) {
+			if (!fu_igsc_device_wait_for_reset(self, error))
+				return FALSE;
+		}
+		if (!fu_device_retry_full(FU_DEVICE(self),
+					  fu_igsc_device_reconnect_cb,
+					  200,
+					  300 /* ms */,
+					  NULL,
+					  error))
+			return FALSE;
+		if (!fu_igsc_device_no_update(self, error)) {
+			g_prefix_error(error, "failed to send no-update: ");
+			return FALSE;
+		}
+	}
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_igsc_device_write_firmware(FuDevice *device,
+			      FuFirmware *firmware,
+			      FuProgress *progress,
+			      FwupdInstallFlags flags,
+			      GError **error)
+{
+	FuIgscDevice *self = FU_IGSC_DEVICE(device);
+	g_autoptr(GBytes) fw_info = NULL;
+	g_autoptr(GBytes) fw_payload = NULL;
+
+	/* get image, and install on ourself */
+	fw_info =
+	    fu_firmware_get_image_by_idx_bytes(firmware, FU_IFWI_FPT_FIRMWARE_IDX_INFO, error);
+	if (fw_info == NULL)
+		return FALSE;
+	fw_payload =
+	    fu_firmware_get_image_by_idx_bytes(firmware, FU_IFWI_FPT_FIRMWARE_IDX_FWIM, error);
+	if (fw_payload == NULL)
+		return FALSE;
+	return fu_igsc_device_write_blob(self,
+					 GSC_FWU_HECI_PAYLOAD_TYPE_GFX_FW,
+					 fw_info,
+					 fw_payload,
+					 progress,
+					 error);
+}
+
+static gboolean
+fu_igsc_device_set_pci_power_policy(FuIgscDevice *self, const gchar *val, GError **error)
+{
+	g_autoptr(FuUdevDevice) parent = NULL;
+
+	/* get PCI parent */
+	parent = fu_udev_device_get_parent_with_subsystem(FU_UDEV_DEVICE(self), "pci");
+	if (parent == NULL) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED, "no PCI parent");
+		return FALSE;
+	}
+	return fu_udev_device_write_sysfs(parent, "power/control", val, error);
+}
+
+static gboolean
+fu_igsc_device_prepare(FuDevice *device,
+		       FuProgress *progress,
+		       FwupdInstallFlags flags,
+		       GError **error)
+{
+	FuIgscDevice *self = FU_IGSC_DEVICE(device);
+	return fu_igsc_device_set_pci_power_policy(self, "on", error);
+}
+
+static gboolean
+fu_igsc_device_cleanup(FuDevice *device,
+		       FuProgress *progress,
+		       FwupdInstallFlags flags,
+		       GError **error)
+{
+	FuIgscDevice *self = FU_IGSC_DEVICE(device);
+	return fu_igsc_device_set_pci_power_policy(self, "auto", error);
+}
+
+static void
+fu_igsc_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 1, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 88, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 1, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 10, "reload");
+}
+
+static void
+fu_igsc_device_init(FuIgscDevice *self)
+{
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
```

Comment:
is this true?

Reply:
I'm getting hardware next week, and so I can find out. :)

Reply:
Nope, removed.